### PR TITLE
`spring cloud` - update sdk to 2023-03-01-preview

### PR DESCRIPTION
**Pull request by ms-henglu**
There's a change introduced in the `2023-03-01-preview`, that the default build service won't be created automatically.

This PR introduces a new field `default_build_service_enabled=true` to create the build service by default.

image

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -26,7 +26,7 @@ require (
 	github.com/rickb777/date v1.12.5-0.20200422084442-6300e543c4d9
 	github.com/sergi/go-diff v1.2.0
 	github.com/tombuildsstuff/giovanni v0.20.0
-	github.com/tombuildsstuff/kermit v0.20230331.1120327
+	github.com/tombuildsstuff/kermit v0.20230412.1161912
```
Comment (tombuildsstuff) [inline go.mod:29]: could we split the SDK update out into another PR?

**Review by tombuildsstuff — CHANGES_REQUESTED** [review]
hey @ms-henglu 

I've taken a look through and left a few comments inline, if we can address those then we should be able to take another look.

Thanks!

File: internal/services/springcloud/spring_cloud_service_resource.go [hunk 2]
```diff
@@ -413,6 +414,17 @@ func resourceSpringCloudServiceCreate(d *pluginsdk.ResourceData, meta interface{
...
+		buildResource := appplatform.BuildService{
+			Properties: &appplatform.BuildServiceProperties{},
+		}
+		buildServiceCreateFuture, err := buildServiceClient.CreateOrUpdate(ctx, id.ResourceGroup, id.SpringName, "default", buildResource)
+		if err != nil {
+			return fmt.Errorf("creating build service %s: %+v", id, err)
+		}
+		if err := buildServiceCreateFuture.WaitForCompletionRef(ctx, client.Client); err != nil {
+			return fmt.Errorf("waiting for creation build service of %s: %+v", id, err)
+		}
```
[the remark this answers is not in the document]
Comment (ms-henglu) [reply internal/services/springcloud/spring_cloud_service_resource.go:426]: Not, it can only be created with name=default.

File: go.mod (hunk 1 above)
Comment (ms-henglu) [reply go.mod:29]: No, we can't split the SDK update only, because it introduces a behavioral breaking change, that the default build service won't be created by the service automatically.

**Comment by ms-henglu** [comment]
Hi @tombuildsstuff ,

Thank you for the code review.

This SDK introduces a behavioral breaking change that the default build service won't be created automatically. And we have 
`spring_cloud_service#build_agent_pool_size` (which manages the default buildService/agentPool resource) and `azurerm_spring_cloud_builder`(which manages buildService/builders resource) depending on the default build service.

Is it okay to separate the `buildService` as a separate resource and deprecate the `spring_cloud_service#build_agent_pool_size` field?

But this would be a breaking change. WDYT?

**Comment by ms-henglu** [comment]
Hi @tombuildsstuff ,

I've updated this PR, now it create the build service by default, and it also exposes `default_build_service_enabled` field, which allows user to disable the default build service. Would you please take another look? Thanks!

File: internal/services/springcloud/spring_cloud_service_resource.go [hunk 3]
```diff
@@ -581,6 +603,22 @@ func resourceSpringCloudServiceRead(d *pluginsdk.ResourceData, meta interface{})
 	if utils.ResponseWasNotFound(serviceRegistry.Response) {
 		serviceRegistryEnabled = false
 	}
+
+	defaultBuildServiceEnabled := true
+	if resp.Sku != nil && resp.Sku.Name != nil && *resp.Sku.Name == "E0" {
```
Comment (katbyte) [inline internal/services/springcloud/spring_cloud_service_resource.go:608]: why are we checking the sku here?

**Review by katbyte — CHANGES_REQUESTED** [review]
We should maintain the existing behaviour users expect, and create the build service by default

File: internal/services/springcloud/spring_cloud_service_resource.go [hunk 4]
```diff
@@ -581,6 +603,22 @@ func resourceSpringCloudServiceRead(d *pluginsdk.ResourceData, meta interface{})
...
+			if !utils.ResponseWasNotFound(defaultBuildService.Response) {
+				return fmt.Errorf("retrieving default build service of %s: %+v", id, err)
+			}
+			defaultBuildServiceEnabled = false
+		}
+		if utils.ResponseWasNotFound(defaultBuildService.Response) {
+			defaultBuildServiceEnabled = false
+		}
+	}
+	d.Set("default_build_service_enabled", defaultBuildServiceEnabled)
```
Comment (tombuildsstuff) [inline internal/services/springcloud/spring_cloud_service_resource.go:620]: why are we doing this only for the SKU E0? Shouldn't we create this always

File: internal/services/springcloud/spring_cloud_service_resource.go [hunk 5]
```diff
@@ -87,6 +87,13 @@ func resourceSpringCloudService() *pluginsdk.Resource {
 				}, false),
 			},
 
+			"default_build_service_enabled": {
+				Type:     pluginsdk.TypeBool,
+				Optional: true,
+				Default:  true,
+				ForceNew: true,
+			},
```
Comment (tombuildsstuff) [inline internal/services/springcloud/spring_cloud_service_resource.go:95]: can we remove this feature flag? this should probably be created by default

File: internal/services/springcloud/spring_cloud_service_resource.go (hunk 3 above)
Comment (ms-henglu) [reply internal/services/springcloud/spring_cloud_service_resource.go:608]: Build service is only supported in enterprise sku.

File: internal/services/springcloud/spring_cloud_service_resource.go [hunk 6]
```diff
@@ -412,6 +420,19 @@ func resourceSpringCloudServiceCreate(d *pluginsdk.ResourceData, meta interface{
 		}
 	}
 
+	if defaultBuildServiceEnabled := d.Get("default_build_service_enabled").(bool); defaultBuildServiceEnabled && skuName == "E0" {
```
[the remark this answers is not in the document]
Comment (ms-henglu) [reply internal/services/springcloud/spring_cloud_service_resource.go:423]: Build service is only supported in enterprise sku.

File: internal/services/springcloud/spring_cloud_service_resource.go (hunk 4 above)
Comment (ms-henglu) [reply internal/services/springcloud/spring_cloud_service_resource.go:620]: Build service is only supported in enterprise sku.

File: internal/services/springcloud/spring_cloud_service_resource.go (hunk 5 above)
Comment (ms-henglu) [reply internal/services/springcloud/spring_cloud_service_resource.go:95]: Sure, I'll remove this flag in this PR.

**Comment by ms-henglu** [comment]
Hi @katbyte , @tombuildsstuff ,

Thanks for the code review!

I've updated this PR to remove `default_build_service_enabled ` flag as suggested, and since the build service is only supported in enterprise sku, so we need to check the sku_name to prevent bad request.
